### ISSUE TITLE
[jvm-packages] ArrayIndexOutOfBoundsException in XGBoosterGetModelRaw at the end of training on Spark

### ISSUE THREAD

**Issue by evgri243** [dmlc/xgboost#3710]
Training is failing with ArrayIndexOutOfBoundsException somewhere next to the end of training (estimated by execution timing).

Environment and versions:

- 0.8 on Databricks 4.1 ML Beta (includes Apache Spark 2.3.0, Scala 2.11)
- 0.8.3 on Databricks 4.2 (includes Apache Spark 2.3.1, Scala 2.11)
- 1 Master and 2 Worker nodes: 140.0 GB Memory, 20 Cores

After some data preparation on R on master, we are passing data to Scala runtime through temp tables. Final dataset size is about a couple of gigs with 183 columns in total.

Training code for 0.8:
```
import ml.dmlc.xgboost4j.scala.spark.XGBoost

val xgbParam = Map("booster" -> "gbtree",
                   "objective" -> "binary:logistic",
                   "eta" -> 0.001,  
                   "gamma" -> 4,   
                   "max_depth" -> 20,
                   "min_child_weigh" -> 10,
                   "nthread" -> 20,
                   "subsample" -> 0.8,
                   "colsample_bytree" -> 0.7,
                   "scale_pos_weight" -> 36,
                   "max_delta_step" -> 2)

val s = System.nanoTime
val xgbClassifier = XGBoost.trainWithDataFrame(training_ml, xgbParam, nrounds, numWorkers)
println("time: " + (System.nanoTime-s)/(1e9*60) + "s")
```

Almost the same code used for 0.8.3:
```
import ml.dmlc.xgboost4j.scala.spark.XGBoostClassifier

val xgbParam = Map("booster" -> "gbtree",
                   "objective" -> "binary:logistic",
                   "eta" -> 0.001,  
                   "gamma" -> 4,   
                  "max_depth" -> 20,
                   "min_child_weigh" -> 10,
                   "subsample" -> 0.8,
                   "colsample_bytree" -> 0.7,
                   "scale_pos_weight" -> 36,
                   "max_delta_step" -> 2)


val xgbClassifier = new XGBoostClassifier(xgbParam)
                       .setFeaturesCol("features")
                       .setLabelCol("label")
                       .setNumRound(10000)
                       .setNumWorkers(numWorkers)
                       .setNthread(20)

val s = System.nanoTime
xgbClassifier.fit(training_ml)
println("time: " + (System.nanoTime-s)/(1e9*60) + "s")
```

Both fails with the following error:

> org.apache.spark.SparkException: Job aborted due to stage failure: Task 0 in stage 6.0 failed 4 times, most recent failure: Lost task 0.3 in stage 6.0 (TID 207, 10.139.64.11, executor 1): java.lang.ArrayIndexOutOfBoundsException
	at ml.dmlc.xgboost4j.java.XGBoostJNI.XGBoosterGetModelRaw(Native Method)
	at ml.dmlc.xgboost4j.java.Booster.toByteArray(Booster.java:435)
	at ml.dmlc.xgboost4j.java.Booster.writeObject(Booster.java:496)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at java.io.ObjectStreamClass.invokeWriteObject(ObjectStreamClass.java:1128)
	at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1496)
	at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1432)
	at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1178)
	at java.io.ObjectOutputStream.defaultWriteFields(ObjectOutputStream.java:1548)
	at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1509)
	at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1432)
	at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1178)
	at java.io.ObjectOutputStream.defaultWriteFields(ObjectOutputStream.java:1548)
	at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1509)
	at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1432)
	at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1178)
	at java.io.ObjectOutputStream.writeArray(ObjectOutputStream.java:1378)
	at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1174)
	at java.io.ObjectOutputStream.writeObject(ObjectOutputStream.java:348)
	at org.apache.spark.serializer.JavaSerializationStream.writeObject(JavaSerializer.scala:43)
	at org.apache.spark.serializer.JavaSerializerInstance.serialize(JavaSerializer.scala:100)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:397)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)

Driver log is omitted for clarity.

**_Final notes:_**

- The same code works fine for 50, 100 and 1000 _numRounds_ and fails on 10000.
- It also works on a dataset with 159 columns, but fails here with 183.
- Spark cluster started with spark.task.cpus=20, but it looks like having no effect as the same error is hit on 40 1-threaded workers.

Any ideas on a way to fix it?

**Comment by hcho3** [comment]
Adding `[jvm-packages]` so that contributors to JVM packages will notice them.

**Comment by CodingCat** [comment]
We never released any version of 0.8.3...the next one should be 0.81...ask Databricks if it’s some customized version
